rfc/text/advanced/ap_authentication_scram.md

```diff
@@ -0,0 +1,387 @@
+             IfNull(CBindData, "", UnBase64(CBindData))
+
+ClientFinalNoProof = "c=" + Base64(CBindInput) + "," +
+                     "r=" + AUTHENTICATE.Extra.nonce
+
+AuthMessage = ClientFirstBare + "," + ServerFirst + "," +
+              ClientFinalNoProof
+```
+
+###### ClientProof
+
+`ClientProof` is computed by the client during the authentication exchange as follows:
+
+{align="left"}
+```javascript
+SaltedPassword  = KDF(Normalize(password), salt, params...)
+ClientKey       = HMAC(SaltedPassword, "Client Key")
+StoredKey       = SHA256(ClientKey)
+ClientSignature = HMAC(StoredKey, AuthMessage)
+ClientProof     = ClientKey XOR ClientSignature
+```
+
+The `ClientProof` is then sent to the server, base64-encoded, via the `AUTHENTICATE.Signature` argument.
+
+The server verifies the `ClientProof` by computing the `RecoveredStoredKey` and comparing it to the actual `StoredKey`:
+
+{align="left"}
+```javascript
+ClientSignature    = HMAC(StoredKey, AuthMessage)
+RecoveredClientKey = ClientSignature XOR ReceivedClientProof
+RecoveredStoredKey = SHA256(RecoveredClientKey)
+```
+
+Note that the client MAY cache the `ClientKey` and `StoredKey` (or just `SaltedPassword`) to avoid having to perform the expensive KDF computation for every authentication exchange. Storing these values securely on the client is outside the scope of this document.
+
+###### ServerSignature
+
+`ServerSignature` is computed by the server during the authentication exchange as follows:
+
+{align="left"}
+```javascript
+ServerSignature = HMAC(ServerKey, AuthMessage)
+```
+
+The `ServerSignature` is then sent to the client, base64-encoded, via the `WELCOME.Details.authextra.verifier` attribute.
+
+The client verifies the `ServerSignature` by computing it and comparing it with the `ServerSignature` sent by the server:
+
+{align="left"}
+```javascript
+ServerKey       = HMAC(SaltedPassword, "Server Key")
+ServerSignature = HMAC(ServerKey, AuthMessage)
+```
+
+##### Key Derivation Functions
+
+SCRAM uses a password-based key derivation function (KDF) to hash user passwords. WAMP-SCRAM supports both [Argon2](https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2/) and [PBKDF2](https://tools.ietf.org/html/rfc2898) as the KDF. Argon2 is recommended because of its memory hardness and resistance against GPU hardware. PBKDF2, which does not feature memory hardness, is also supported for applications that are required to use primitives currently approved by cryptographic standards.
+
+The following table maps the `CHALLENGE.Details.kdf` type string to the corresponding KDF.
+
+KDF type string | Function
+--------------- | --------
+`"argon2id13"` | Argon2id variant of Argon2, version 1.3
+`"pbkdf2"`      | PBKDF2
+
+WAMP-SCRAM client implementations SHALL support **all** of the above KDFs. During authentication, there is no "negotiation" of the KDF, and the client MUST use the same KDF than the one used to create the keys stored in the authentication database.
```

Comment:
and servers? I guess both client and router implementations SHOULD support all KDFs? plus: ".. and SHOULD prefer argon2id13 to pbkdf2"?

Reply:
Isn't it more important for servers to support "all" KDFs? Then client implementations can choose to support only one (i.e. could be "SHOULD support all KDFs" for clients and "MUST support all KDFs" for servers).

Reply:
You have to keep in mind that *user registration* is a whole other process that is separate from *user authentication*. After the user is registered and the hashed keys are stored in the authentication database, the client no longer has any choice whether to use argon2id13 or pbkdf2 for that particular user. The client *must* authenticate using the same hash algorithm that was used for user registration.

For example, one router can choose to only store argon2id13-derived keys, and another router chooses to store only pbkdf2-derived keys. For a WAMP client implementation to universally support all routers, the client implementation would need to support both argon2id13 and pbkdf2.

If user registration was part of the spec, then I could see how it would be desirable to require that routers support both argon2id13 and pbkdf2. But since we don't impose any requirements for user registration, I don't see the need to impose both argon2id13 and pbkdf2 for router implementations.

Reply:
I see -- so you're imposing the "SHALL" on client implementations so that interop is more likely -- but isn't it just as likely that some client decides "we'll only do `argon2id13`" as that some router will decide that? I mean, I could see that being "policy" for an application just as easily as a router...?

"The thing that's doing registrations" would be where the KDF is negotiated (if at all), so you'd just fail to be able to register if you were using a router that said "`argon2id13`-only" and your client only did `pbkdf2` (or vice versa obviously).

Also worth noting that I presume the registration-process would involve the server/router telling the client: KDF, cost, memory, salt and then the client hashing its password and sending the appropriate results back to the server -- maybe it's worth mentioning this briefly somewhere in the spec?

Reply:
I now realize that requiring the client to support both algorithms, but not the router, is going to be confusing to implementors/adopters, so I'm willing to change it so that both client/server are required to support both algorithms. If there's an application/site policy of only using a certain KDF, then they can configure their router accordingly.

I someone wants to later change the spec it so that it's not required to support both algorithms, they can raise an issue and make their case.

> Also worth noting that I presume the registration-process would involve the server/router telling the client: KDF, cost, memory, salt and then the client hashing its password and sending the appropriate results back to the server -- maybe it's worth mentioning this briefly somewhere in the spec?

The registration process does not necessarily involve the client and server talking to each other. The router administrator could directly add the user and assign a known password by configuring a file or manipulating a database.

If a router implementation decides to provide a client API for registering a user, the hashing could be done on the server side if the cleartext password is sent directly by the client (presumably over a secure encrypted transport). Alternatively, the hashing could be done client-side if the server first sends it the algorithm/salt/cost parameters. There are so many possible permutations that I'm not comfortable specifying exactly how it should be done.

Given that the server does not need the cleartext password, it would be preferable from a security standpoint to have it hashed client side during registration that involves. But that would place a burden on the registration client on knowing how to compute the hashes and stored keys.

Reply:
I don't see the point of using SCRAM if the server/router ever sees the plaintext password, personally. The client already knows how to compute the correct pieces, since it has to do that to authenticate in the first place (i.e. it has to produce `SaltedPassword` and `StoredKey` already; "registration" would merely be sending those along to the server with the parameters used).

Anyway, that said I agree with you that there's lots of ways to do it so maybe suggesting any at all is counter-productive. "Outside the scope" is fine :)

A very simple registration method would have the client produce the values and send them all to the server -- the server could reject it if e.g. the parameters were too simple, etc. (The server could compute `ServerKey` itself from `SaltedPassword`). So, yes, again this doesn't involve the server talking "first" and sending params.

Reply:
I am +1 on: "both clients and routers SHOULD implement both KDF functions". there is no point in "MUST" - I am simply going to ignore that as an implementor and can claim the app code just disabled or doesn't use the respective unimplemente KDF;) We should leave that to implementors - as with implementing WAMP-SCRAM at all ..

Reply:
Hmm... at first you guys wanted both to be mandatory to avoid negotiating the KDF, and now you want it to be up to the implementation. Is this change of mind due to better understanding that there's no KDF negotiation involved during user authentication, but only possibly during user registration?

I'm fine with making it "both clients and routers SHOULD implement both KDF functions", as long as we cease flip-flopping on this matter. :smile:

Reply:
right;) trying to nalil it, unify: both KDFs should be possible, but IMO we should take an opinionated stand here in favor of argon:

* argon: both client/router implementations MUST support it
* pbkdf2: both client/router implementations SHOULD support it

so this leaves open the option to run pbkdf2 in environments where this is necessary or wanted for whatever reasons, but allows implementations to only implement argon, and makes it the default choice